clippy_lints/src/boxed_void.rs

```diff
@@ -0,0 +1,50 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::path_def_id;
+use clippy_utils::ty::is_c_void;
+use rustc_hir::{Expr, ExprKind, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::RawPtr;
+use rustc_middle::ty::TypeAndMut;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks if we're passing a `c_void` raw pointer to `Box::from_raw(_)`
+    ///
+    /// ### Why is this bad?
+    /// This is a known pitfall of FFI, particularly with some C or C++ APIs.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::ffi::c_void;
+    /// let ptr = Box::into_raw(Box::new(42usize)) as *mut c_void;
+    /// let _ = unsafe { Box::from_raw(ptr) };
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::ffi::c_void;
+    /// let ptr = Box::into_raw(Box::new(42usize)) as *mut c_void;
+    /// let _ = unsafe { Box::from_raw(ptr as *mut usize) };
+    /// ```
+    #[clippy::version = "1.66.0"]
+    pub BOXED_VOID,
```

Comment:
I would maybe make the name more specific, maybe `RAW_BOX_FROM_CVOID_PTR` since `void` and `unit` are often used interchangeably, and we're not talking about the `()` type here

Reply:
SGTM, but I would drop the `C` from the same, so just `RAW_BOX_FROM_VOID_PTR`

Reply:
Renamed the lint as suggested.